### PR TITLE
Add wasi_snapshot and make it default for wasi

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

File: std/assembly/bindings/wasi_snapshot.ts [hunk 1]
```diff
@@ -869,22 +869,26 @@ export namespace errno {
...
+@unmanaged export class event {
   /** User-provided value that got attached to `subscription#userdata`. */
   userdata: userdata;
   /** If non-zero, an error that occurred while processing the subscription request. */
   error: errno;
-  /* The type of the event that occurred. */
+  /** The type of the event that occurred. */
   type: eventtype;
+  /** The contents of the event. */
+  u: event_fd_readwrite;
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi_snapshot.ts]: Iirc this is a union, but `u` would point somewhere here. Idea previously was that after `type` there is `__padding0` followed by the extensions performed by `rwevent` or any other concrete event. One would then cast `event` to `rwevent` if `event#type` indicates an `rwevent`, in turn gaining access to the extended data of the union **if** an rwevent.

Comment (MaxGraey) [reply std/assembly/bindings/wasi_snapshot.ts]: Hmm, so I should revert this changes and use previous approach?

Comment (dcodeIO) [reply std/assembly/bindings/wasi_snapshot.ts]: It's fine to update the layout if it has changed ofc, but the new approach with `u` doesn't seem correct as in AS every field of a class type is a pointer, and there is no pointer here. The only way to model a union is with `extends` I think.

**Comment by MaxGraey** [comment]
I preserved `wasi_unstable.ts` but `wasi.ts` now redirect to `wasi_snapshot.ts`. We still need backward compatibility. Also `node 13.3.0` it seems implement `wasi_unstable` version

**Comment by dcodeIO** [comment]
Hmm, [bytecodealliance/wasi](https://github.com/bytecodealliance/wasi) references a `wasi_snapshot_preview1` version. Is there a difference to `wasi_snapshot`?

**Comment by MaxGraey** [comment]
It's `wasi_snapshot_preview0`. Need check which changes introduced in `snapshot1`...

**Comment by MaxGraey** [comment]
Ahh, sorry it's `snapshot_preview1`.

**Comment by dcodeIO** [comment]
I see. Now I'm wondering how we can make this pluggable with a `--use` option. For instance if there is a wasi program that is intended to run in engines implementing different standards. 

Maybe we can do

```ts
// bindings/wasi.ts
export * as unstable from "./wasi_unstable.ts";
export * as snapshot_preview1 from "./wasi_snapshot.ts";
```

and then utilize `--use wasi=bindings/wasi/snapshot_preview1` or something like that, with the default being the latest bindings we have?

**Comment by MaxGraey** [comment]
Or just `--use wasi=snapshot_preview1` or `--use wasi=unstable`

**Comment by dcodeIO** [comment]
Thanks! I guess it's fine to look into the mechanism mentioned above later.

**Pull request merged**